core/identity-hub-core/src/main/java/org/eclipse/edc/identityhub/core/CredentialQueryResolverImpl.java

```diff
@@ -0,0 +1,150 @@
+import org.eclipse.edc.identityhub.spi.store.CredentialStore;
+import org.eclipse.edc.identityhub.spi.store.model.VerifiableCredentialResource;
+import org.eclipse.edc.spi.query.Criterion;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.result.AbstractResult;
+import org.eclipse.edc.spi.result.Result;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import static org.eclipse.edc.spi.result.Result.failure;
+import static org.eclipse.edc.spi.result.Result.success;
+
+
+public class CredentialQueryResolverImpl implements CredentialQueryResolver {
+
+    private final CredentialStore credentialStore;
+    private final ScopeToCriterionTransformer scopeTransformer;
+
+    public CredentialQueryResolverImpl(CredentialStore credentialStore, ScopeToCriterionTransformer scopeTransformer) {
+        this.credentialStore = credentialStore;
+        this.scopeTransformer = scopeTransformer;
+    }
+
+    @Override
+    public QueryResult query(PresentationQuery query, List<String> issuerScopes) {
+        if (query.getPresentationDefinition() != null) {
+            throw new UnsupportedOperationException("Querying with a DIF Presentation Exchange definition is not yet supported.");
+        }
+        if (query.getScopes().isEmpty()) {
+            return QueryResult.noScopeFound("Invalid query: must contain at least one scope.");
+        }
+
+        // check that all prover scopes are valid
+        var proverScopeResult = parseScopes(query.getScopes());
+        if (proverScopeResult.failed()) return QueryResult.invalidScope(proverScopeResult.getFailureMessages());
+
+        // check that all issuer scopes are valid
+        var issuerScopeResult = parseScopes(issuerScopes);
+        if (issuerScopeResult.failed()) return QueryResult.invalidScope(issuerScopeResult.getFailureMessages());
+
+        // query storage for requested credentials
+        var queryspec = convertToQuerySpec(proverScopeResult.getContent());
+        var res = credentialStore.query(queryspec);
+        if (res.failed()) {
+            return QueryResult.storageFailure(res.getFailureMessages());
+        }
+
+        // the credentials requested by the other party
+        var requestedCredentials = res.getContent().toList();
+
+        // check that prover scope is not wider than issuer scope
+        var issuerQuery = convertToQuerySpec(issuerScopeResult.getContent());
+        var predicate = issuerQuery.getFilterExpression().stream()
+                .map(c -> credentialsPredicate(c.getOperandRight().toString()))
+                .reduce(Predicate::or)
+                .orElse(x -> false);
+
+        // now narrow down the requested credentials to only contain allowed creds
+        var allowedCredentials = requestedCredentials.stream().filter(predicate).toList();
+
+        var isValidQuery = validateResults(new ArrayList<>(requestedCredentials), new ArrayList<>(allowedCredentials));
```

Comment:
I think this line could be replaced by:
```java
var isValidQuery = requestedCredentials.stream().filter(predicate).count() == requestedCredentials.size();
```

the behavior will be the same (and tests still pass), and `validateResult` can be deleted